### ISSUE TITLE
Add __slots__ to pl.Expr and pl.LazyFrame

### ISSUE THREAD

**Issue by Object905** [pola-rs/polars#13198]
### Description

Profiled count of created pl.Expr and pl.LazyFrame in my use case (graphQL api on top of polars). 
Found that one of the heaviest queries creates pl.Expr around 3k times and pl.LazyFrame around 1k times.

Tried to "monkey patch" adding __slots__ on them directly in .venv and found that it seems to be working ok.

Will follow up with PR later.

**Comment by ritchie46** [comment]
What will that do?

**Comment by mcrumiller** [comment]
`__slots__` improves the performance of class attribute access, sometimes by quite a bit. I tried playing with this in dataframe/series a while back and ran into issues with namespace registration, since using `__slots__` means you can't dynamically add attributes to a class. To get around this, I had to create a separate namespace dict and override `__getattr__` to detect namespace retrieval, and in doing so removed the performance benefit of using `__slots__`.

**Comment by Object905** [comment]
Well, then registration of namespace would be better handled by using [descriptors](https://docs.python.org/3/howto/descriptor.html). 

Like this. But will experiment on this once I get to PR.

```python
class Namespace:
    def __get__(self, obj, objtype=None):
        return "FooNamespace"


class Foo:
    __slots__ = ("_bar",)


Foo.namespace = Namespace()

instance = Foo()
print(instance.namespace)  # prints FooNamespace
print(Foo.namespace) # prints FooNamespace too
```

**Comment by Object905** [comment]
Having __slots__ only means that you can't add attributes to instances, not class itself .  Class object would still have __dict__.

**Comment by Object905** [comment]
[NameSpace ](https://github.com/pola-rs/polars/blob/5bca54b0148b01e08efd88fe40fa31d385990070/py-polars/polars/api.py#L34C1-L34C16) class that wraps plugins is already is a descriptor. And it seems like it's registered on classes in [_create_namespace](https://github.com/pola-rs/polars/blob/5bca54b0148b01e08efd88fe40fa31d385990070/py-polars/polars/api.py#L50C5-L50C22), and that should not interact with `__dict__ `of instances in any way.
So, given that it goes to `Expr.__dict__` I don't see why you would need to define `__getattr__` on Expr.

@mcrumiller Am I missing something here?

**Comment by mcrumiller** [comment]
@Object905 if it works, go for it! I may have done something incorrect, and this was a while ago so I don't recall the details. Are you talking about adding `_expr` and `_ldf` to the `__slots__`?

**Comment by Object905** [comment]
Yes. Builtin namespaces, like str, list, etc also good candidates.

**Comment by mcrumiller** [comment]
@knl I was just about to make a feature request for an empty `attr` dictionary.
